src/sentry_metrics.c

```diff
@@ -0,0 +1,1004 @@
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+
+#define ROLLUP_IN_SECONDS 10
+#define MAX_TOTAL_WEIGHT 100000
+
+#ifdef SENTRY_UNITTEST
+#define FLUSHER_SLEEP_TIME_SEC 0
+#else
+#define FLUSHER_SLEEP_TIME_SEC 5
+#endif
+
+
+static bool g_aggregator_initialized = false;
+static sentry_metrics_aggregator_t g_aggregator = { 0 };
+static sentry_mutex_t g_aggregator_lock = SENTRY__MUTEX_INIT;
+
+static int32_t g_total_buckets_weight = 0;
+static bool g_is_flush_scheduled = false;
+
+static bool g_is_closed = false;
+
+static sentry_bgworker_t *g_metrics_bgw = NULL;
+
+static sentry_value_t
+sentry__metrics_type_to_string(sentry_metric_type_t type)
+{
+    switch (type) {
+    case SENTRY_METRIC_COUNTER:
+        return sentry_value_new_string("counter");
+    case SENTRY_METRIC_DISTRIBUTION:
+        return sentry_value_new_string("distribution");
+    case SENTRY_METRIC_GAUGE:
+        return sentry_value_new_string("gauge");
+    case SENTRY_METRIC_SET:
+        return sentry_value_new_string("set");
+    default:
+        assert(!"invalid metric type");
+        return sentry_value_new_null();
+    }
+}
+
+static const char *
+sentry__metrics_type_to_statsd(sentry_metric_type_t type)
+{
+    switch (type) {
+    case SENTRY_METRIC_COUNTER:
+        return "c";
+    case SENTRY_METRIC_DISTRIBUTION:
+        return "d";
+    case SENTRY_METRIC_GAUGE:
+        return "g";
+    case SENTRY_METRIC_SET:
+        return "s";
+    default:
+        assert(!"invalid metric type");
+        return "invalid";
+    }
+}
+
+static sentry_metric_type_t
+sentry__metrics_type_from_string(sentry_value_t type)
+{
+    const char *type_str = sentry_value_as_string(type);
+
+    if (sentry__string_eq(type_str, "counter")) {
+        return SENTRY_METRIC_COUNTER;
+    } else if (sentry__string_eq(type_str, "distribution")) {
+        return SENTRY_METRIC_DISTRIBUTION;
+    } else if (sentry__string_eq(type_str, "gauge")) {
+        return SENTRY_METRIC_GAUGE;
+    } else if (sentry__string_eq(type_str, "set")) {
+        return SENTRY_METRIC_SET;
+    } else {
+        assert(!"invalid metric type");
+        return -1;
+    }
+}
+
+int
+has_name_pattern_match(char c)
+{
+    return isalnum(c) || c == '_' || c == '\\' || c == '-' || c == '.';
+}
+
+int
+has_unit_pattern_match(char c)
+{
+    return isalnum(c) || c == '_';
+}
+
+int
+has_tag_key_pattern_match(char c)
+{
+    return isalnum(c) || c == '_' || c == '\\' || c == '-' || c == '.'
+        || c == '/';
+}
+
+char *
+sentry__metrics_sanitize(const char *original, const char *replacement,
+    int (*pattern_match_func)(char c))
+{
+    sentry_stringbuilder_t sb;
+    sentry__stringbuilder_init(&sb);
+
+    const unsigned char *ptr = (const unsigned char *)original;
+    for (; *ptr; ptr++) {
+        if (pattern_match_func(*ptr)) {
+            sentry__stringbuilder_append_char(&sb, *ptr);
+        } else {
+            sentry__stringbuilder_append(&sb, replacement);
+        }
+    }
```

Comment:
The version here is safe concerning UTF-8 multibyte characters. But it produces one replacement character per byte instead of per character. To fix this, we must skip over the continuation bytes in the sequence like so:

```suggestion
    while (*ptr) {
        if (pattern_match_func(*ptr)) {
            sentry__stringbuilder_append_char(&sb, *ptr);
            ptr++;
        } else {
            sentry__stringbuilder_append(&sb, replacement);
            ptr++;

            // At this point, the last `ptr` value was either some replaced
            // ASCII or the start of a multi-byte sequence, which means `ptr`
            // points to the next character or the second byte of a multi-byte
            // sequence. If it is the latter, we must skip over all bytes in the
            // sequence so we only replace the whole character once.
            // Continuation bytes have the most significant bits set to `10`.
            while ((*ptr & 0xC0) == 0x80) {
                ptr++;
            }
        }
    }
```